enhancements/dns/plugins.md

```diff
@@ -0,0 +1,332 @@
+2. How are API changes handled for cluster downgrades?
+3. Should IPv6 references for `Nameservers` be removed until OpenShift adds IPv6 support?
+4. Is the cluster domain (i.e. cluster.local) an invalid `ForwardPlugin` domain?
+5. Should the number of `ForwardPlugins` be restricted due to Gap 4?
+
+## Summary
+
+DNS name resolution for services and pods is currently provided by an instance of [CoreDNS](https://coredns.io) that
+runs on each node in the cluster. The [cluster-dns-operator](https://github.com/openshift/cluster-dns-operator) manages
+the CoreDNS configuration with a statically-defined ConfigMap:
+
+```bash
+$ oc get configmap/dns-default -n openshift-dns -o yaml
+apiVersion: v1
+data:
+  Corefile: |
+    .:5353 {
+        errors
+        health
+        kubernetes cluster.local in-addr.arpa ip6.arpa {
+            pods insecure
+            upstream
+            fallthrough in-addr.arpa ip6.arpa
+        }
+        prometheus :9153
+        forward . /etc/resolv.conf
+        cache 30
+        reload
+    }
+kind: ConfigMap
+metadata:
+  labels:
+    dns.operator.openshift.io/owning-dns: default
+  name: dns-default
+  namespace: openshift-dns
+  <SNIP>
+```
+
+Once CoreDNS starts and has parsed the configuration, it runs servers. Each server is defined by the zones it serves and
+a listening port. In the above configuration, CoreDNS starts one server that manages all zones and listens on port 5353.
+Each server has its own plugin chain represented within the server block stanza (i.e. `forward`). This proposal will
+generate the ConfigMap with a `forward` plugin configuration based on user-provided values of type `ForwardPlugin`.
+
+## Motivation
+
+CoreDNS is responsible for resolving pod and service names for the cluster domain (i.e. `cluster.local`). Otherwise,
+CoreDNS proxies the request to a resolver identified by `/etc/resolv.conf` on the corresponding node. Although this provides a consistent and reliable approach for name
+resolution, it restricts how cluster operators manage DNS name queries.
+
+### Goals
+
+1. A well defined API for managing CoreDNS plugins.
+2. The ability to configure the CoreDNS forwarding plugin.
+2. A minimal API surface that can be expanded to support future DNS forwarding and plugin use cases.
+
+### Non-Goals
+
+1. Support every possible DNS forwarding use case.
+2. Configure or manage external DNS providers.
+3. Provide name query forwarding for other cluster services (i.e. container runtime).
+4. Manage what plugins get loaded (i.e. plugin.cfg).
+
+## Proposal
+
+`PluginConfig` defines one or more CoreDNS plugins that can be configured. If a plugin is not defined, it will use a
+default configuration provided by the cluster-dns-operator. A plugin is represented as a slice when multiple instances
+of the plugin can be expressed within the DNS configuration.
+
+```go
+type PluginConfig struct {
+    ForwardPlugins []ForwardPlugin `json:"forwardPlugin"`
```

Comment:
The Go field is plural (`ForwardPlugins`), but the json tag is singular (`forwardPlugin`).